File: Python/0-1 Knapsack Problems/min_sum_difference.py
def subset(n, arr, s):
    dp = [[0 for _ in range(s+1)] for _ in range(n+1)]

    for i in range(n+1):
        for j in range(s+1):
            if(i==0):
                dp[i][j] = False
            elif(j==0):
                dp[i][j] = True
            elif(arr[i-1]<=j):
                dp[i][j] = dp[i-1][j] or dp[i-1][j-arr[i-1]]
            else:
                dp[i][j] = dp[i-1][j]
    return dp[-1]

def sumdiff(n, arr):
    mini = 0
    maxi = sum(arr)
    ans = float('inf')
    dp = subset(n, arr, maxi)
    #print(dp)
    for i in range(mini, (maxi+2)//2):
        if(dp[i]):
            ans = min(ans, maxi-2*i)
    return ans
    


test = int(input())
for i in range(test):
    n = int(input())
    arr = list(map(int, input().split()))
    print(sumdiff(n, arr))

File: C++/0-1 Knapsack Problems/min_sum_difference.cpp
#include<bits/stdc++.h>
#include<numeric>
using namespace std;

int sumdiff(int n, int arr[]){
    int mini = 0;
    int maxi = accumulate(arr, arr+n, maxi);
    int ans = INT_MAX;
    bool dp[n+1][maxi+1];

    for(int i=0; i<=n; i++){
        for(int j=0; j<=maxi; j++){
            if(i==0){
                dp[i][j] = false;
            } else if(j==0){
                dp[i][j] = true;
            } else if(arr[i-1]<=j){
                dp[i][j] = dp[i-1][j] || dp[i-1][j-arr[i-1]];
            } else{
                dp[i][j] = dp[i-1][j];
            }
        }
    }
    for(int i=mini; i<=(maxi)/2; i++){
        //cout<<ans<<endl;
        if(dp[n][i] == 1){
            int res = maxi - 2*i;
            if(ans>res){
                ans = res;
            }
        }
    }
    return ans;
}

int main() {
    int test, n;
    cin>>test;
    while(test--){
        cin>>n;
        int arr[n];
        for(int i=0; i<n; i++){
            cin>>arr[i];
        }
        cout<<sumdiff(n, arr)<<'\n';
    }
}
